### DISCUSSION TITLE
Docker in WSL vs. Docker Desktop on Host

### DISCUSSION THREAD

**Discussion by nico-loeber** [ddev/ddev#3784] [General]
Hello,
as @rfay already noted a few times, ddev might change the docs to suggest installing docker inside WSL instead of using Docker Desktop on the host. I would like to understand and discuss the reasons for this.

The idea of WSL in Windows is a fusion of the advantages of both operation systems and as much interoperability as possible.
This starts with the Windows Terminal that allows you to easily switch between a windows and a linux terminal, the possibility to access both file systems from each other, as well as many other features that allow interoperability between both systems.
Software like VSCode, PHPStorm adapt this idea and allow you to benefit from WSL features even they run in windows.
Docker Desktop does the same as it allows you to run your containers in the WSL Backend while you can control them from your windows host or any wsl distribution. It also provides a GUI to manage your images, volumes and provides even a dockerhub integration.

Installing Docker directly in Linux might cause some some issues in my opinion. You can not manage your images from your windows environment and you do not have a GUI. Also windows won't be aware about the installed docker and there won't be any entries in the windows hosts file. The hosts file of your WSL distributions will automatically be overwriten by your windows one, if you do not disable this behavior.
This might make networking and accessing your containers more difficult.
Also you need to reinstall your WSL from time to time to be able to use new features or for an easy dist upgrade. In this case I just backup my ssh config and I push my latest project changes, the I simply remove the distribution and install the new version. I don't want to have any software I have to configure installed in my WSL. This might be a personal flavor of course.

I do see some benefits in using docker directly inside wsl. You do have the "original" docker and you do not need to care about licenses. Also I prefer administrating linux environment over windows. Installing and configuring stuff is more simple. But this shouldn't be a reason, since you already made the decission to prefer windows if you come to the question we just discuss.
For sure there are more benefits I'm not thinking of, so please tell me. :)

I would suggest that ddev should be independent from either docker installation method. The documentation should just state that docker needs to be available and should reference related external documentation.
There shouldn't be a "default" way, even though I personally prefer docker desktop as it fits better the idea of WSL, a seamless integration that connect the best of both worlds.

**Comment by rfay** [comment]
Great conversation!

The key benefits of installing docker inside the container are:
* Stability: Linux docker is generally rock-solid, whereas Docker Desktop frequently breaks on new releases.
* Open-Source: The linux docker is fully open-source and has a clearer future than Docker Desktop.
* Licensing: Docker Desktop is a licensed product, and requires a per-seat licensing fee in many companies.

In general, the filesystem and networking issues.are the same between Docker Desktop and the docker-inside-WSL2 approach.

**Comment by nico-loeber** [comment]
Okay, thank you for clarifying!

I can understand that stability is an important point. I didn't experience any problems with docker desktop in the last year. But i only use it for basic tasks like running ddev, or testing gitlab or elastic in a container.

I think it is important to consider your target group. 

Linux Fanboys
Most of them use Ubuntu, many experiment with smaller distributions. They are capable of administrating their system via CLI. Open source software and Licenses are an important topic for them, not only because of the money, but because of their personal principles. Individuals or small companies belong to this group.

Mac Users
They love their OS and everything apple produces. Many are familiar with the CLI, but the system administration is mostly done by them self via GUI. Money is no topic and they would probably pay for good software. Individuals and all types of companies belong to this group.

Windows Users
Windows Users just "use" their OS and don't like to administrate it. I don't know many web developers who are Windows fans but they use it since it has some advantages, though. Nearly all administrative tasks are done via GUI or by a company administrator / GPO. If open source would be an important topic they wouldn't use windows. For developing Linux related software, WSL is used to mitigate the disadvantages of the windows file system and command line. Some individuals and all types of companies belong to this group.

I apologize if anyone feels offended by this stigma.

Lets compare the advantages and disadvantages of both solutions:

| Docker Desktop                                 | Docker (inside WSL)                    |
|------------------------------------------------|--------------------------------------|
| Free for private use and small companies       | Open-Source                          |
| Automatic updates or via package manager       | updates via package manager          |
| quite stable                                   | rock-solid                           |
| Available on all WSL distributions and Windows | Installation per distribution        |
| GUI & CLI                                           | CLI only                             |
| Includes docker compose                        | Compose plugin needs to be installed |
| Manages networking for host                    | Manual configuration needed          |
| Available for Windows, MacOS and Linux         | Linux only                           |

In my Opinion using a docker installation via command line would fit the profile of a Linux and maybe aMac user but not that of a windows user.
Windows users like a good GUI, don't care much about the license. Also installing docker desktop requires no configuration for use in Windows or WSL and can be pre installed by a company administrator easily.

So I still think any docker installation should be supported by ddev and none should be suggested.

**Reply by rfay** [reply]
I do think you underestimate Windows users. What CMS community are you associated with? In the TYPO3 community people have been flocking to Windows and WSL2 and they're very experienced and capable people. And they've been using WSl2 with ddev in very advanced ways. 

Thanks for the great comparison. 

Note that your statement about "Includes docker-compose" isn't relevant to DDEV users, as ddev provides its own docker-compose, as I'm sure you know. So for DDEV nothing needs to be installed in that regard. 

IMO the installation path for WSL2 is way too hard no matter where you install Docker. It's hard with Docker Desktop, it's the same amount of hard with docker-inside-WSL2. I'd just like to find ways to make it much easier for everybody.

**Reply by rfay** [reply]
I do think you underestimate Windows users. What CMS community are you associated with? In the TYPO3 community people have been flocking to Windows and WSL2 and they're very experienced and capable people. And they've been using WSl2 with ddev in very advanced ways. 

Thanks for the great comparison. 

Note that your statement about "Includes docker-compose" isn't relevant to DDEV users, as ddev provides its own docker-compose, as I'm sure you know. So for DDEV nothing needs to be installed in that regard. 

IMO the installation path for WSL2 is way too hard no matter where you install Docker. It's hard with Docker Desktop, it's the same amount of hard with docker-inside-WSL2. I'd just like to find ways to make it much easier for everybody.

**Reply by nico-loeber** [reply]
I AM Windows User, that's why this topic is important to me. I am working in a company with a wide range of peoples windows skills. Including project managers, designers, frontend developers, backend developers and windows administrators. Sorry again if my post was offensive in any way.
And asking for my "CMS community"... we develop our own enterprise CMS System. So there is not much of a community. ;)

**Comment by nico-loeber** [comment]
@rfay I understand your points, nevertheless I would like to write some concept paper suggesting an alternative installation approach for dddev in Windows WSL, if you allow me to. It will include Docker Desktop as a mainstay. I know about the licensing problems for larger companies, and I can't provide a solution for that, but as I already mentioned, companies that want to use ddev and have high enough revenue to need to pay for docker desktop will most likely just pay for a license, the pricing is quite fair.

**Reply by rfay** [reply]
I'd love great ideas about how to do a ddev installation that's easier! That would be fantastic.

There's not really a whole lot of difference between using Docker Desktop and docker-inside-wsl2, so I imagine everything you suggest would be useful for both approaches.

**Comment by nico-loeber** [comment]
@rfay, here you can find my concept on how to setup DDEV inside WSL based on the interop feature.
This will increase interoperability and will reduce the setup complexity and even solves a problem I stumbled across.

https://gist.github.com/nico-loeber/6d6e8261827c5e4efa4ccf5b5cd04140

Feedback is welcome.
